Add commands to weave files and preview them in Atom

pfitzseb opened the pull request:
If Juno (and Weave.jl) s installed you can now `weave` arbitrary files and preview them in Atom.

This implements a HTML preview pane and makes use of https://github.com/izuzak/atom-pdf-view for previewing PDFs.

![image](https://user-images.githubusercontent.com/6735977/53697281-55979480-3dd0-11e9-805e-4b215565bbfa.png)

The previews auto-reload when the `*.html`/`*.pdf` changes.

Possible improvements:
1. Allow setting a different path for generating output.
2. Or, more generally, allow specifying arbitrary kwargs for `weave`.

Not sure how much either of those are used in practice.

mpastell commented:
Thanks! This looks great! I think keyword arguments and separate ouput directories are used quite much, but they can always be added later. 

Would it make sense to transfer this repo to Junolab organization as this brings even tighter Juno intergration?

pfitzseb commented:
I'm not entirely sure how to best handle kwargs, because it only makes sense to set them per project/file. So I see two options:
1. Show a UI for every new file to allow setting kwargs (and cache them, of course).
2. Allow putting all kwargs into the weave header as settings.

> Would it make sense to transfer this repo to Junolab organization as this brings even tighter Juno intergration?

Your call. I'd be happy helping maintain this package wherever it is ;)

mpastell commented:
I think allowing to set keyword arguments in the header makes sense. I'll try to add that,

mpastell commented:
I think transferring the repo makes sense. Is it OK if I just transfer the ownership to Junolab?

pfitzseb commented:
One other thing that would be very helpful is if `weave` returned the location of the output file.

> I think transferring the repo makes sense. Is it OK if I just transfer the ownership to Junolab?

Yeah, sure!

mpastell commented:
It seems that I'd need permission to create repos in JunoLab to make the transfer: https://help.github.com/en/articles/transferring-a-repository#repository-transfers-and-organizations

> **Repository transfers and organizations**
> To transfer repositories to an organization, you must have repository creation permissions in the receiving organization. If organization owners have disabled repository creation by organization members, only organization owners can transfer repositories out of or into the organization.
> Once a repository is transferred to an organization, the organization's default repository permission settings and default membership privileges will apply to the transferred repository.

mpastell commented:
Pdf preview works great, but I get this error for html and the preview appears, but on top of REPL instead of it's own tab.

[Enter steps to reproduce:]

1. ...
2. ...

**Atom**: 1.34.0 x64
**Electron**: 2.0.16
**OS**: Ubuntu 18.04.2
**Thrown From**: [language-weave](https://github.com/mpastell/language-weave) package 0.5.0


### Stack Trace

Uncaught Error: Cannot call reloadIgnoringCache because the webContents is unavailable. The WebView must be attached to the DOM and the dom-ready event emitted before this method can be called.

```
At /snap/atom/222/usr/share/atom/resources/electron.asar/renderer/web-view/web-view.js:390

Error: Cannot call reloadIgnoringCache because the webContents is unavailable. The WebView must be attached to the DOM and the dom-ready event emitted before this method can be called.
    at HTMLElement.reloadIgnoringCache (/snap/atom/222/usr/share/atom/resources/electron.asar/renderer/web-view/web-view.js:390:15)
    at HTMLPreview.reloadWebview (/packages/language-weave/lib/html-preview.js:45:26)
    at /packages/language-weave/lib/html-preview.js:34:26)
    at emitTwo (events.js:126:13)
    at FSWatcher.emit (events.js:214:7)
    at FSEvent.FSWatcher._handle.onchange (fs.js:1368:12)
```

### Commands

```
     -4:13.3.0 command-palette:toggle (ul.list-inline.tab-bar.inset-panel)
     -4:10.5.0 core:move-down (input.hidden-input)
 11x -4:09.2.0 core:backspace (input.hidden-input)
     -4:02.9.0 core:confirm (input.hidden-input)
     -4:02.9.0 settings-view:open (ul.list-inline.tab-bar.inset-panel)
     -3:48.3.0 weave:weave-to-html (input.hidden-input)
```

### Non-Core Packages

```
indent-detective 0.3.2 
ink 0.9.14 
julia-client 0.7.13 
language-julia 0.18.0 
language-weave 0.5.0 
latex-completions 0.3.6 
pdf-view 0.71.0 
tool-bar 1.1.11 
uber-juno 0.2.0 
```

mpastell commented:
On Windows I get this error from both pdf and html buttons:

`Base.Meta.ParseError("invalid escape sequence")`

mpastell commented on lib/main.js:
```diff
@@ -24,8 +28,126 @@ export function activate () {
+    outpath = edpath.split('.')
+    outpath.pop()
+    outpath = outpath.join('.') + '.' + format
+
```
Adding `edpath = edpath.replace(/\\/g, "/")` fixes the error I got on Windows.

pfitzseb commented:
Ok, I think I've fixed the problem you had with the HTML view (I hope?) and incorporated your windows fix. Also sent you an invite to JunoLab, so I think you should be able to transfer the repo now.

mpastell commented on lib/html-preview.js:
```diff
@@ -0,0 +1,83 @@
+'use babel'
+
```
I needed to add `const path = require('path')` here to make html preview work.

mpastell commented:
It seems I still can't transfer the repo, I get an error that I'm not allowed to create a repository in JunoLab. Can you rights to do it temporarily? 

I needed to make one more small fix to get html preview working (see code comment). 

It works really well now, really a great addition to Weave workflow!

pfitzseb commented:
Ok, you should have permissions to create repos now.
Also fixed the missing `path` import, so this should be good to go now?

mpastell commented:
Thanks! Everything seems to work, I'll publish a release and transfer the repo.

mpastell commented:
Repository is now transferred, thanks! 

FYI: `weave` now returns the document name after weaving and you can set the document options in header.
